### PR TITLE
Disallow `ispath(::Integer)` and similar for a bunch of other methods (`isdir`, `isfile`, ...)

### LINKED ISSUE #51710: ispath returning false positive result for the given argument
julia> ispath("/tmp")
true

julia> ispath("1")
false

julia> ispath(1)
true

image

### PR DESCRIPTION
**Pull request by jakobnissen**
Before, `ispath(p)` forwarded to `ispath(stat(p))`, and `stat` took an integer, presumably interpreting as a OS file descriptor.
But, having `ispath(1)` return true is strange and grating.

Closes #51710 

NOTE TO REVIEWERS: Was the old behaviour `ispath(1) === true` intended? If so, this is a breaking change, and not a bugfix.

### CODE REVIEW

**Comment by Seelengrab** [comment]
I'm surprised any of the functions from that loop just take an `Int`. The documentation for all of these certainly only seems to suggest that this is an actual path pointing to some file. So might be good to just disallow it for all of these?

**Comment by StefanKarpinski** [comment]
Restricting this is definitely good but this feels like not enough. We should systematically limit all of these path functions, not just ispath.

**Comment by jakobnissen** [comment]
Okay, I've now gone through the docstrings of each of these functions in the list. For those that say the argument should be a path, I've changed the function to require an `AbstractString`.
However, some functions directly mention compatibility to `stat` in their docstring, e.g. `uperm` which states:
> For allowed arguments, see `stat`.

Notably, `stat` itself doesn't mention that it can take integers. The relevant part of the docstring is:
> stat(file)
>
> Return a structure whose fields contain information about the file.

However, I fear it might be breaking to remove this functionality from `stat`.
I don't have enough experience to understand when you'd want to use an integer (file ID?) to `stat` and related functions, so this is purely out of precaution. Maybe it's fine to also limit `stat` (and `uperm` etc) to take `AbstractString`. If so, tell me and I'll update this PR.

**Comment by jakobnissen** [comment]
I'm marking this with `minor change`, because all the functions that no longer take an integer was only ever documented to take a path. However, whether it actually is a minor change or a breaking change in practice, I'm not sure.
Perhaps we could do a PkgEval run?

**Comment by vtjnash** [comment]
Marking for triage to merge this

**Comment by LilithHafner** [comment]
Triage says yes (though we still need code review)

**Comment by iamed2** [comment]
It's possible that FilePathsBase.jl should react to this, I made an issue there: https://github.com/rofinn/FilePathsBase.jl/issues/170

It's likely (but unconfirmed) the impact there is small and limited to undocumented functions that were previously working.

**Comment by vtjnash** [comment]
FilePathsBase.jl seems to be a good argument against doing this. Perhaps we just add an explicit deny-list of types (e.g. subtypes of `Number`)? In particular, I know that package used to subtype AbstractString so that patterns like this worked, but then eventually concluded that wasn't correct anyways: https://github.com/rofinn/FilePathsBase.jl/blob/39587a14bfb19fc05e016b460f5cf75c27be7876/docs/src/faq.md?plain=1#L23-L26

**Comment by jakobnissen** [comment]
I don't think it's unreasonable to expect a library for working with paths to implement some of the methods themselves - and @rofinn seems open to updating FilePathsBase.
I'd rather have sensible semantics in Base (i.e. an integer is not a path), and then have the author of path libraries decide explicitly how to implement their functions.

**Comment by vtjnash** [comment]
In general, that seems it would lead to APIs that are not composable, and not typical of how interfaces are written. That is just the nature of "duck-typing". But I realize the issue here isn't that an error happens later when passing something that doesn't conform to the interface, but rather that something unexpectedly works, because it does conform to the interface.

Story time: long ago, there was a minor push to make file descriptors use the RawFD type instead of an untyped Int, because it would have avoided this particular problem. But we weren't quite certain that a RawFD needed to be different from an Int so the work didn't get too far. It became far more relevant on Windows, since Int is ambiguous when attempting to treat it as a file handle what kind of file handle it is attempting to represent (is it a HANDLE or an HINSTANCE or an MSVC fd, for example). I don't think we ever got far enough on the work to decide if it would be a breaking change to entirely stop treating `Int` as an equivalent (but not equal) alias for `RawFD`.

**Comment by rofinn** [comment]
FWIW, I doubt this will break anything on the FilePathsBase.jl side. `Base.ispath(::AbstractPath)` is simply a wrapper to `FilePathsBase.exists(::AbstractPath)` for easier integration with existing code that works on strings. Since the proposal is just to let a `MethodError` be thrown, I don't see any issue. 

Just to be safe I've run all the package tests on this branch, and nothing breaks. I've also looked at `AWSS3.S3Path`, which just implements the `exists` method.

**Comment by vtjnash** [comment]
It could be more so a conflict with other local changes such as https://github.com/JuliaLang/julia/pull/53377, where it may be reasonable to pass through all `stat` function helpers, since it provides many of them already. Just as an example where the current untyped API can make things simpler

File: base/stat.jl [hunk 1]
```diff
@@ -439,29 +439,37 @@ operm(st::StatStruct) = UInt8((filemode(st)     ) & 0x7)
...
-    :uperm,
-    :gperm,
-    :operm,
-    :filemode,
-    :filesize,
-    :mtime,
-    :ctime,
 ]
-    @eval ($f)(path...)  = ($f)(stat(path...))
+    @eval ($f)(path::AbstractString) = ($f)(stat(path))
```
Comment (tecosaur) [inline base/stat.jl:469]: Should this keep the `...`?
```julia
@eval ($f)(path::AbstractString...) = ($f)(stat(path...))
```

Comment (Seelengrab) [reply base/stat.jl:469]: Probably also take at least one path; `isfile()` doesn't really make sense.

**Comment by tecosaur** [comment]
@jakobnissen given all the comments here (and triage approval), it _looks_ like this should be good to merge? I think it could be worth getting your thoughts on the "explicit MethodError with ::Int" alternative approach though.

File: base/stat.jl (hunk 1 above)
Comment (tecosaur) [reply base/stat.jl:469]: Mmmm, `(path0::AbstractString, paths::AbstractString...)` or similar seems like a good idea.

**Comment by jakobnissen** [comment]
My thoughts are that I don't think it's a good idea. Functions like `isfile` should operate on paths, that's it. They should not operate on dicts, sets, lists, `Nothing`s, or just any old random value. This just makes semantic sense. Since we (unfortunately) don't have a path type in Julia, it'll have to be strings instead.

I think if people want to invent new data types like to represent paths, they'll have to add new methods themselves. I don't think that's unreasonable to expect.
It's possible I misunderstood the motivation for allowing `::Any` in these methods, but I really don't see the point. Integers aren't paths. Neither are linked lists, or matrices, etc.

**Comment by LilithHafner** [comment]
@nanosoldier `runtests()`

**Comment by vtjnash** [comment]
> My thoughts are that I don't think it's a good idea

Whether you like it or not, Julia is a duck-typed language, and what you describe does not fit with that expectation of the way the language, as currently defined and implemented. What you seem to be suggesting is that Julia should have interfaces, which is entirely reasonable, but just not what it has now.

**Comment by nanosoldier** [comment]
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/51711#issuecomment-2069672719) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/e017c11_vs_b5bfd83/report.html) is available.

**Comment by LilithHafner** [comment]
At least two failures look real:

https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/e017c11_vs_b5bfd83/MatrixDepot.primary.log:
```
 MethodError: no method matching isfile(::IOStream)
  The function `isfile` exists, but no method is defined for this combination of argument types
```

https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/e017c11_vs_b5bfd83/FileJockey.primary.log:
```
 Expression: entry("sandbox/") |> is(DirEntry)
  MethodError: no method matching isdir(::DirEntry)
  The function `isdir` exists, but no method is defined for this combination of argument types.
```

**Comment by jakobnissen** [comment]
What the right way to get this done? Triage has approved this as a minor change, but there are PkgEval failiures. I can make PRs to failing packages if needed.

**Comment by LilithHafner** [comment]
Triage's yes was before we had nanosoldier results and I don't think FilePathsBase came up in the triage discussion. Maybe worth chatting again at triage?

**Comment by rofinn** [comment]
Sorry, why does triage need to discuss FilePathsBase? I don't see anything specific to that package in the failures, just a couple of precompilation issues and some calls to `isfile(::IOStream)` that would need to be deprecated?

**Comment by LilithHafner** [comment]
FilePathsBase is skipped by Nanosoldier, I imagine it would have failed.

**Comment by rofinn** [comment]
Interesting. Everything passed when I ran those tests on this branch before, but I'm not familiar with the nanosoldier workflow or runtime environment.

**Comment by LilithHafner** [comment]
Oh, if you tested locally and it passed then it passes. I was just thinking that some generic fallbacks via stat might need to be re-implemented if this merges.

**Comment by mbauman** [comment]
There is an alternative: we could `@deprecate` the integer `stat` method in favor of the explicit `RawFD` one — `RawFD` is exported and documented for exactly this purpose.

That’d be nonbreaking and simpler.

**Comment by fingolfin** [comment]
Do we still need/wants this once PR #55080 is merged?

**Comment by oscardssmith** [comment]
yes. #55080 is the prereq (by making it so that people don't end up with Ints when they wanted FDs, but this PR is needed to make sure that people don't try to do FD operations with random Ints they have.

**Comment by LilithHafner** [comment]
`ispath(::Integer)` is deprecated since #55250 and I think the way to disallow it is #54855.

**Comment by oscardssmith** [comment]
Should we also explicitly deprecate `isdir`, `isfile` etc so the error message is better?

**Pull request closed without merging**